plugin-server/src/utils/db/db.ts

```diff
@@ -529,6 +540,123 @@ export class DB {
+            ]
+        )
+    }
+
+    private getPropertyLastUpdatedAtDateTimeOrEpoch(person: Person, key: string): DateTime {
+        const lookup = person.properties_last_updated_at[key]
+        if (lookup) {
+            return DateTime.fromISO(lookup)
+        }
+        return DateTime.fromMillis(0)
+    }
+
+    private getPropertiesLastOperationOrSet(person: Person, key: string): PersonPropertyUpdateOperation {
+        if (person.properties_last_operation === null || !(key in person.properties_last_operation)) {
+            return PersonPropertyUpdateOperation.Set
+        }
+        return person.properties_last_operation[key]
+    }
+
+    private updatePropertiesLastOperation(person: Person, key: string, value: PersonPropertyUpdateOperation) {
+        // to get around the type check
+        if (person.properties_last_operation === null) {
+            person.properties_last_operation = {}
+        }
+        person.properties_last_operation[key] = value
+    }
+
+    private updatePersonPropertiesLocal(
+        person: Person,
+        properties: Properties,
+        propertiesOnce: Properties,
+        timestamp: DateTime
+    ): boolean {
+        // updates the person & returns true/false if anything was updated
+        let updatedSomething = false
+        Object.entries(propertiesOnce).forEach(([key, value]) => {
+            if (
+                !(key in person.properties) ||
+                (this.getPropertiesLastOperationOrSet(person, key) == PersonPropertyUpdateOperation.SetOnce &&
+                    this.getPropertyLastUpdatedAtDateTimeOrEpoch(person, key) > timestamp)
+            ) {
+                updatedSomething = true
+                person.properties[key] = value
+                this.updatePropertiesLastOperation(person, key, PersonPropertyUpdateOperation.SetOnce)
+                person.properties_last_updated_at[key] = timestamp.toISO()
+            }
+        })
+        // note that if the key appears twice we override it with set value here
+        Object.entries(properties).forEach(([key, value]) => {
+            if (
+                !(key in person.properties) ||
+                this.getPropertiesLastOperationOrSet(person, key) == PersonPropertyUpdateOperation.SetOnce ||
+                this.getPropertyLastUpdatedAtDateTimeOrEpoch(person, key) < timestamp
+            ) {
+                updatedSomething = true
+                person.properties[key] = value
+                this.updatePropertiesLastOperation(person, key, PersonPropertyUpdateOperation.Set)
+                person.properties_last_updated_at[key] = timestamp.toISO()
+            }
+        })
+        return updatedSomething
+    }
+
+    public async updatePersonProperties(
+        teamId: number,
+        distinctId: string,
+        properties: Properties,
+        propertiesOnce: Properties,
+        timestamp: DateTime
+    ): Promise<void> {
+        if (properties.length === 0 && propertiesOnce.length === 0) {
```

Comment:
I'm quite sure the `Properties` type is not an array, meaning `.length` won't work here.

Reply:
But because JavaScript, this doesn't throw an error, instead it evaluates to `undefined && undefined`

Reply:
will add a test for this